scripts/flights2.py

```diff
@@ -0,0 +1,727 @@
+        fp = Path(source)
+        msg = f"Reading specs from {fp.as_posix()!r}"
+        logger.info(msg)
+        mapping = tomllib.loads(fp.read_text("utf-8"))
+        if specs_array := mapping.get("specs"):
+            return cls(
+                specs=[Spec.from_dict(spec) for spec in specs_array],
+                input_dir=input_dir or mapping["input_dir"],
+                output_dir=output_dir or mapping["output_dir"],
+            )
+        msg = (
+            f"Expected to find an array of tables keyed to `'specs'`, but got\n"
+            f"{mapping!r}"
+        )
+        raise TypeError(msg)
+
+    def __iter__(self) -> Iterator[Spec]:
+        yield from self.specs
+
+    @property
+    def ranges(self) -> pl.LazyFrame:
+        return pl.select(pl.concat(spec.range.monthly for spec in self), eager=False)
+
+    @property
+    def _required_stems(self) -> set[str]:
+        date = col("date")
+        return set(
+            self.ranges.select(
+                date.dt.year().alias("year"), date.dt.month().alias("month")
+            )
+            .unique()
+            .select(_file_stem_source("year", "month"))
+            .collect()
+            .to_series()
+            .to_list()
+        )
+
+    @property
+    def _existing_stems(self) -> set[str]:
+        return {_without_suffixes(fp.name) for fp in app.input_dir.glob(PATTERN_GZIP)}
+
+    @property
+    def missing_stems(self) -> set[str]:
+        missing = self._required_stems - self._existing_stems
+        if n := len(missing):
+            msg = f"Missing {n} sources"
+            logger.info(msg)
+            if n >= 5:
+                logger.warning("Downloads may exceed 100MB")
+            if n >= 11:
+                logger.warning("Total number of rows will exceed 5_000_000")
+        return missing
+
+    async def _download_sources_async(self, names: Iterable[str], /) -> list[Path]:
+        """Request, write missing data."""
+        session = niquests.AsyncSession(base_url=ROUTE_ZIP)
+        aws = (_request_async(session, name) for name in names)
+        buffers = await asyncio.gather(*aws)
+        writes = (_write_rezip_async(self.input_dir, buf) for buf in buffers)
+        return await asyncio.gather(*writes)
+
+    def download_sources(self) -> None:
+        """
+        Ensure all required source data is saved to ``self.input_dir``.
+
+        Any month(s) that are missing will be requested from `transtats`_.
+
+        .. _transtats:
+            https://www.transtats.bts.gov
+        """
+        logger.info("Detecting required sources ...")
+        if missing := self.missing_stems:
+            asyncio.run(self._download_sources_async(missing))
+            logger.info("Successfully downloaded all missing sources.")
+        else:
+            logger.info("Sources already downloaded.")
+
+    def scan_sources(self) -> SourceMap:
+        logger.info("Scanning dependencies ...")
+        for spec in self:
```

Comment:
### TODO
- [ ] Needs doc

Reply:
Resolved in (https://github.com/vega/vega-datasets/pull/645/commits/99169b6ba5afa77df90228519280cc4c7ff23305)